developer_guides/virtualisation/virtualisation.rst

```diff
@@ -0,0 +1,255 @@
+.. _virtualisation:
+
+SOF VirtIO design
+#################
+
+.. contents::
+
+1. General information
+**********************
+
+VirtIO is a platform-agnostic virtualisation standard, managed by OASIS
+https://www.oasis-open.org/committees/tc_home.php?wg_abbrev=virtio . It defines
+a design and a protocol for implementing device IO in VMs.
+
+When Linux is running in a VM, multiple drivers are already available, using
+virtio: networking, storage, etc. Those drivers use the virtio and virtqueue
+APIs to communicate with the host or, potentially, with other guests.
+
+The host part depends on the used virtualisation solution. Some of them
+implement the VirtIO API themselves, then using their method of choice to talk
+to the whatever host they are running on. A typical Linux virtualisation
+implementation uses QEMU + KVM to run guest instances. In this case there are
+two ways to implement a VirtIO host: in the user space and in the kernel. If
+implemented in the user space, it is managed by QEMU. In the kernel space VirtIO
+host drivers use the Linux vhost API.
+
+It is also worth to note, that VirtIO guest drivers reside in respective driver
+subdirectories in the Linux kernel, e.g. drivers/net, drivers/block etc. On the
+contrary, all vhost drivers are located under drivers/vhost. That directory also
+contains vhost.c and vhost.h which are used by all vhost drivers.
+
+The SOF virtualisation solution also implements a vhost driver.
+
+2. SOF VirtIO design
+********************
+
+The principal point of the SOF VirtIO design is using the IPC API, that is also
+used by the native SOF driver to communicate with the SOF code, running on the
+audio DSP. We use the same IPC structures and we add 3 more VirtIO-specific IPC
+messages, which are never sent to the DSP, so the firmware doesn't need to
+support them. For the most part IPCs from guests are forwarded one-to-one by the
+host driver to the DSP.
+
+2.1. Communication channels
+---------------------------
+
+The goal of the SOF virtualisation is to implement low latency audio
+virtualisation, which means avoiding copying of audio data and stream position
+updates. In this case only one VirtQueue would be required for IPC message
```

Comment:
low latency also means avoid context changes as much as possible.

Reply:
@plbossart is it worth mentioning that? That isn't really VirtIO specific, that's generic

Reply:
The goal of the SOF virtualisation is to support differentiation in guest audio capabilities and implement low latency audio....